### ISSUE TITLE
Improve performance

### ISSUE THREAD

**Issue by merelinguist** [blitz-js/superjson#25]
Recent benchmarks against a similar library, [devalue](https://github.com/Rich-Harris/devalue), read as follows:

```
superjson output: 179 bytes
devalue output: 116 bytes

() => superjson.stringify(obj)
1000000 iterations in 21005ms

() => devalue(obj)
1000000 iterations in 8168ms

() => superjson.parse(superjson_serialized)
1000000 iterations in 6487ms

() => eval(`(${devalue_serialized})`)
1000000 iterations in 2290ms
```

I’m not so concerned about the output size difference, which I think for now is negligible, but more about the speed. Obviously we’re making different design choices here which necessarily are probably slower, but I think it’d be good to see what improvements can be made in this area.

0.021005ms is pretty quick, though!

Thanks to Rich for setting up the comparison: https://github.com/Rich-Harris/superjson-and-devalue

**Comment by Skn0tt** [comment]
Yeah, we'll definitely need to look into performance. I think there's a lot of low-hanging fruits to be reaped by changing stuff like `[...path, head]` to `path.push(head)` if that's possible.

**Comment by Skn0tt** [comment]
We should tackle #10 before this to have some objective measurement I guess :D

**Comment by merelinguist** [comment]
Yep! Was going to just comment on that but thought I’d set up a new issue.

**Comment by flybayer** [comment]
One thing to keep in mind, is that on the server string<>object serialization & deserialization is currently handled by Next. So we won't use `superjson.stringify`, but we will on the client.

That said, we _could_ use `superjson.stringify` on the server and bypass the Next `res.json()` helper.

**Comment by Skn0tt** [comment]
> I’m not so concerned about the output size difference, which I think for now is negligible, but more about the speed.

In my opinion, we should focus the other way around. UX-wise, compute cycles are cheap but network usage is expensive.

**Comment by flybayer** [comment]
So we already have a slow perf report from someone using Blitz, albeit to load 20k items in one request :)

https://github.com/blitz-js/blitz/issues/955

**Comment by Skn0tt** [comment]
Uh-oh ... I guess we‘ll need to look into adding a bigger benchmark case ^^

**Comment by sannajammeh** [comment]
This still the case?
